craig/atomic_transactions

CraigLawrence opened the pull request:
1) The existing `perl-Data-Chronicle` only provides access to Redis' `get` and `set` method.
This was previously acceptable because our primary use for it, `perl-App-Config-Chronicle`, stored the app configuration as one monolithic entry.
In the new `perl-App-Config-Chronicle` we'd like to store each app config item as an individual entry, which means we'd like access to Redis' `mget` and `mset` methods so that we can read/write sets of individual entries atomically.
I have added `mget` and `mset` methods to this package, and retained the existing `get` and `set` for backwards compatibility and ease of interacting with single entries.

2) I have also add `subscribe` and `unsubscribe` methods in a new package.
This class provides the convenient `publish_on_set` but no direct way to subscribe to the publications. A user would need to call from outside the class `$writer->cache_writer->subscribe($category . '::' . $name, $coderef)`. This also shows poor encapsulation because it requires the user to construct the underlying key from `$category` and `$name` which should stay internal to this package.

CraigLawrence committed:
created new mset mget subs

CraigLawrence committed:
added returns;

CraigLawrence committed:
suppress publication flag

CraigLawrence committed:
added sub unsub calls

CraigLawrence committed:
added returns

CraigLawrence committed:
added setnx method

CraigLawrence committed:
removed pub and archive from setnx

CraigLawrence committed:
tidy

CraigLawrence committed:
removed unused

CraigLawrence committed:
added test for setnx

CraigLawrence committed:
added sub unsub tests

CraigLawrence committed:
added comment

CraigLawrence committed:
removed setnx

zakame commented on lib/Data/Chronicle/Reader.pm:
```diff
@@ -126,13 +126,30 @@ sub get {
-        my $cached_data = $self->cache_reader->get($key);
-        return JSON::MaybeXS->new->decode(decode_utf8($cached_data)) if defined $cached_data;
+        my @cached_data = $self->cache_reader->mget(@keys);
+        return map { JSON::MaybeXS->new->decode(decode_utf8($_)) if $_ } @cached_data;
```
May be worth using the new `JSON::MaybeUTF8` here now.

zakame commented on lib/Data/Chronicle/Writer.pm:
```diff
@@ -128,38 +128,120 @@ publish "category1::name1" in Redis if C<publish_on_set> is true.
+    my $rec_date     = shift;
+    my $archive      = shift // 1;
+    my $suppress_pub = shift // 0;
+    my $ttl          = shift // $self->ttl;
```
Is there a particular reason why the more standard

```perl
my ($self, $category, $name, $value, $rec_date, $archive, $suppress_pub, $ttl) = @_;
```

is not used here, aside from appearance (the per-line declaration/inits are neat, I admit.)

CraigLawrence replied:
No particular reason besides it being the existing style of the file, and it does look neat.

turnstep replied:
+1 for per-line declare/set at the same time [yes it's neat]

CraigLawrence committed:
switch to new MaybeUTF8 lib

CraigLawrence committed:
rest of changes

CraigLawrence committed:
removed Encode dependency

chylli commented on lib/Data/Chronicle/Writer.pm:
```diff
@@ -128,38 +127,119 @@ publish "category1::name1" in Redis if C<publish_on_set> is true.
+    return 1;
+}
+
+=head2 mset
```
The pod need to be updated.

CraigLawrence committed:
fixed pod

CraigLawrence committed:
updated version and readme

CraigLawrence committed:
reverted ver changes

CraigLawrence committed:
tense consistency

zakame committed:
Format latest Changes

This lets the latest release changes render nicer on MetaCPAN.

zakame committed:
v0.17

        - Replace dbh with DBIx::Connector object
        - Add mget and mset methods
        - Add get_history method
        - Add subscribe and unsubscribe methods
        - Replace JSON::MaybeXS and Encode with JSON::MaybeUTF8

zakame commented:
This is now on CPAN as [0.17](https://metacpan.org/release/BINARY/Data-Chronicle-0.17)

leonerd commented on lib/Data/Chronicle/Reader.pm:
```diff
@@ -126,13 +125,30 @@ sub get {
+    return undef;
+}
+
+=head3 C<< my $data = mget([["category1", "name1"], ["category2", "name2"], ...]) >>
```
FYI: I'm aware this `=head3` line is just matching style with the rest of the module, so don't feel you should change it.

But... it's generally best to have headings be simple method names like

    =head3 mget

and then include a one-line synopsis example right below it to demonstrate the syntax. Many of the HTML generation programs use that heading to generate HTML anchors, so if you stick to just `mget` then links of the form `L<Data::Chronicle::Reader/mget>` should work.

Probably a change to make wholesale across the module at another time.

CraigLawrence replied:
Fixed

leonerd commented on lib/Data/Chronicle/Writer.pm:
```diff
@@ -128,38 +127,119 @@ publish "category1::name1" in Redis if C<publish_on_set> is true.
+    die 'Subscription requires a coderef' if ref $subref ne 'CODE';
+
+    my $key = $self->_generate_key($category, $name);
+    return $self->cache_writer->subscribe($key, $subref);
```
I admit I'm having difficulty working out what type a `cache_writer` would be here - it doesn't seem to be documented in this file what that would be. Is there a better explanation somewhere?

But if it's going to be something Redis-related (perhaps using `Net::Async::Redis`), there's a problem in that a subscribing connection cannot also be used for regular data updates. If subscriptions are required, you'll have to maintain two connections.

CraigLawrence replied:
In our case it's an instance of `RedisDB`.

This is actually an issue that should probably get properly resolved in this package, because I think even the original developer didn't have it completely straight.

https://github.com/binary-com/perl-Data-Chronicle/blob/77853aa1c0bde788885edad2c8507875b9ed8d10/lib/Data/Chronicle/Reader.pm#L131
In the `Reader` package the codes checks if `cache_reader` is a blessed object or not, and treats it accordingly. Where `Writer` happily calls `multi`, `publish`, `set`, and `exec` on `cache_writer` with no such check.

I will remove this ambiguity from the code.

leonerd commented on lib/Data/Chronicle/Writer.pm:
```diff
@@ -128,38 +127,119 @@ publish "category1::name1" in Redis if C<publish_on_set> is true.
+sub unsubscribe {
+    my ($self, $category, $name, $subref) = @_;
+    die 'publish_on_set must be enabled to unsubscribe' unless $self->publish_on_set;
+    die 'Unsubscription requires a coderef' if ref $subref ne 'CODE';
```
Any reason you have to pass the same coderef for unsubscribe? That makes it difficult to easily perform subscription using code like

    $writer->subscribe(category => name => sub { ... });

because now you'll have to have saved that particular coderef to pass it later for unsubscribe.

CraigLawrence replied:
I think I made an error here by looking at the wrong documentation, I will go back and change this. Thank you for also bringing to my attention the issue of Redis connection in the previous comment.

CraigLawrence committed:
created separate package for subscription connections

CraigLawrence committed:
removed ambiguity re cache backend

CraigLawrence committed:
fixed subscriptions and fully implemented hash alternative cache

CraigLawrence committed:
changed subref to callback

CraigLawrence committed:
fixed used of $reader

CraigLawrence committed:
removed hash alternative

CraigLawrence committed:
added rev changes

CraigLawrence committed:
updated docs

CraigLawrence committed:
added docs to changes

leonerd commented on README.md:
```diff
@@ -44,6 +50,12 @@ Given a category, name, and revision returns version of the data the specificied
 If revision 1 is chosen, the previous version of the data will be returned.
 
+* **subscribe** (in Data::Chronicle::Subscriber):
+Given a category, name, and callback assigns the callback to be called when a new value is set for the specificied category and name (if the writer has publish_on_set enabled).
```
Typo "specified"

leonerd commented on README.md:
```diff
@@ -44,6 +50,12 @@ Given a category, name, and revision returns version of the data the specificied
+Given a category, name, and callback assigns the callback to be called when a new value is set for the specificied category and name (if the writer has publish_on_set enabled).
+
+* **unsubscribe** (in Data::Chronicle::Subscriber):
+Given a category, name, clears the callbacks associated with the specificied category and name.
```
again

leonerd commented on lib/Data/Chronicle.pm:
```diff
@@ -49,12 +49,22 @@ MUST be either hash-ref or array-ref.
     $writer->set("category1", "name2", "value2", Date::Utility->new("2016-08-01 00:06:00"));
 
+=head2 L<Data::Chronicle::Writer/mset>
+Given multiple categories, names and values atomically performs the set operation on each corresponding category, name, value set.
```
POD requires a blank line after the `=head2` directive. Also for others below

leonerd commented on lib/Data/Chronicle/Reader.pm:
```diff
@@ -90,10 +66,11 @@ Given a category, name and timestamp returns version of data under "category::na
 
-cache_reader can be an object which has `get` method used to fetch data.
-or it can be a plain hash-ref.
+cache_reader should be an object of RedisDB.
```
More accurately wording: "should be an instance of L\<RedisDB\>" - also using `L<...>` to link to the class docs.

leonerd commented on lib/Data/Chronicle/Subscriber.pm:
```diff
@@ -0,0 +1,168 @@
+
+=item B<Transparent>
+
+This modules hides all the details about distribution, caching, database structure and ... from developer. He only needs to call a method
```
"and ..." seems odd wording in documentation. Could you be more specific?

CraigLawrence replied:
I didn't write that, I assume the original author just meant `etc`.
I'll re-word it.

leonerd commented on lib/Data/Chronicle/Subscriber.pm:
```diff
@@ -0,0 +1,168 @@
+ $chronicle_w->set("log_files", "syslog", $d);
+ 
+ #unsubscribe to changes to syslog
+ $chronicle_s->subscribe("log_files", "syslog"); 
```
Maybe `$chronicle_s->unsubscribe` here?

leonerd commented on lib/Data/Chronicle/Reader.pm:
```diff
@@ -90,10 +66,11 @@ Given a category, name and timestamp returns version of data under "category::na
 
 =cut
 
+=head1 Members
```
These are "methods" in Perl.

leonerd commented on lib/Data/Chronicle/Reader.pm:
```diff
@@ -43,26 +42,6 @@ to save data and another method to retrieve it. All the underlying complexities
-
-=back
-
 =head1 Example
```
I know this was existing code, but could you fix the `=head1` capitalisation to be consistent? It's fairly common style that all `head1` titles be in ALL CAPITALS.

leonerd commented on lib/Data/Chronicle/Subscriber.pm:
```diff
@@ -0,0 +1,168 @@
+
+=back
+
+=head1 Example
```
=head1 EXAMPLE

leonerd commented on lib/Data/Chronicle/Subscriber.pm:
```diff
@@ -0,0 +1,168 @@
+
+=cut
+
+=head1 Members
```
=head1 METHODS

leonerd commented on lib/Data/Chronicle/Subscriber.pm:
```diff
@@ -0,0 +1,168 @@
+
+=head2 cache_subscriber
+
+cache_subscriber should be an object of RedisDB.
```
Same comment as earlier

leonerd commented on lib/Data/Chronicle/Subscriber.pm:
```diff
@@ -0,0 +1,168 @@
+
+=item * Search CPAN
+
+L<http://search.cpan.org/dist/Data-Chronicle/>
```
This looks like standard boilerplate from one of the automated module generators.

search.cpan.org is now no more, and is redirecting to metacpan.org. This should probably link to

    L<https://metacpan.org/release/Data-Chronicle/>

Ditto in any other files.

leonerd commented on lib/Data/Chronicle/Writer.pm:
```diff
@@ -69,19 +69,31 @@ to save data and another method to retrieve it. All the underlying complexities
 
 =cut
 
+=head1 Members
```
Same comments in this file too about

    =head1 METHODS

Also "instance of", use of `L<...>` to refer to other class names

leonerd commented on lib/Data/Chronicle/Writer.pm:
```diff
@@ -128,38 +140,80 @@ publish "category1::name1" in Redis if C<publish_on_set> is true.
+    my $suppress_pub = shift // 0;
+    my $ttl          = shift // $self->ttl;
+
+    $self->_validate_value($_->[2]) foreach @$entries;
```
If there's a lot of values to be set, and one of them fails validation, the error message won't mention the name of the one that failed. This probably makes debugging more difficult. Could you arrange to pass the name too? Maybe

    $self->_validate_value(@$_) foreach @$entries;

then it has access to the category and name too.

leonerd commented on lib/Data/Chronicle/Writer.pm:
```diff
@@ -128,38 +140,80 @@ publish "category1::name1" in Redis if C<publish_on_set> is true.
+    foreach my $entry (@$entries) {
+        my $category = $entry->[0];
+        my $name     = $entry->[1];
+        my $value    = $entry->[2];
```
You could unpack all three with

    my ($category, $name, $value) = @$entry;

leonerd commented on lib/Data/Chronicle/Writer.pm:
```diff
@@ -128,38 +140,80 @@ publish "category1::name1" in Redis if C<publish_on_set> is true.
 }
 
+sub _validate_value {
+    my ($self, $value) = @_;
```
With the above suggestion, this can now become something like

    my ($self, $category, $name, $value) = @_;
    die "Cannot store an undefined value for ${category}::${name} in Chronicle!" ...

CraigLawrence committed:
Paul's style changes

CraigLawrence committed:
review changes

CraigLawrence committed:
syntax fixes

CraigLawrence committed:
fix line endings
